src/runtime-rs/crates/resource/src/cgroups/mod.rs

```diff
@@ -79,78 +80,69 @@ impl CgroupsResource {
-            None
-        };
+        // Init cgroups for sandbox and overhead
+        let (sandbox_cgroup, overhead_cgroup) =
+            Self::new_cgroups(&config).context("new cgroups")?;
 
         // Add the runtime to the VMM sandbox resource controller
-
-        // By adding the runtime process to either the sandbox or overhead controller, we are making
-        // sure that any child process of the runtime (i.e. *all* processes serving a Kata pod)
-        // will initially live in this controller. Depending on the sandbox_cgroup_only settings, we will
-        // then move the vCPU threads between resource controllers.
         let pid = CgroupPid { pid: 0 };
-        if let Some(manager) = overhead_cgroup_manager.as_ref() {
-            manager.add_task_by_tgid(pid).context("add task by tgid")?;
+        if let Some(cgroup) = overhead_cgroup.as_ref() {
+            cgroup
+                .add_task_by_tgid(pid)
+                .context("add runtime to overhead cgroup")?;
         } else {
-            cgroup_manager
+            sandbox_cgroup
                 .add_task_by_tgid(pid)
                 .context("add task by tgid with sandbox only")?;
         }
 
         Ok(Self {
-            cgroup_manager,
+            cgroup_manager: sandbox_cgroup,
             resources: Arc::new(RwLock::new(HashMap::new())),
-            overhead_cgroup_manager,
+            overhead_cgroup_manager: overhead_cgroup,
             cgroup_config: config,
         })
     }
 
-    /// delete will move the running processes in the cgroup_manager and
-    /// overhead_cgroup_manager to the parent and then delete the cgroups.
-    pub async fn delete(&self) -> Result<()> {
-        for cg_pid in self.cgroup_manager.procs() {
-            // For now, we can't guarantee that the process in cgroup_manager does still
-            // exist. Once it exit, we should ignore that error returned by remove_task_by_tgid
-            // to let it go.
-            if let Err(error) = self.cgroup_manager.remove_task_by_tgid(cg_pid) {
-                match error.source() {
-                    Some(err) => match err.downcast_ref::<io::Error>() {
-                        Some(e) => {
-                            if e.raw_os_error() != Some(OS_ERROR_NO_SUCH_PROCESS) {
-                                return Err(error.into());
-                            }
-                        }
-                        None => return Err(error.into()),
-                    },
-                    None => return Err(error.into()),
-                }
+    fn new_cgroups(config: &CgroupConfig) -> Result<(Cgroup, Option<Cgroup>)> {
+        let new_cgroup = |path: &str| -> Result<Cgroup> {
+            let mut cg_builder = CgroupBuilder::new(path);
+            if config.threaded_mode() {
+                // write "+{controller} to cgroup.subtree_control
+                cg_builder = cg_builder.set_specified_controllers(config.threaded_controllers());
+            }
```

Comment:
Why is this necessary OR why is this only done in threaded mode?
I would expect that you want this for all cgroups2 cases, as cgroup controllers are enabled hierarchically (so this acts on the root).

Reply:
We only constraint non-user workload threads with the overhead cgroup, so that the related controllers are cpu and cpuset. Other controllers, e.g. memory, are managed by the pod cgroup at the proc level, and it is not necessary to manage the controllers at the thread level. @yaoyinnan posted a diagram to help explain the hierarchy, you may refer to [1].

1: https://github.com/kata-containers/kata-containers/pull/5491#issuecomment-1911609331

Reply:
> Why is this necessary OR why is this only done in threaded mode?

If we disable the overhead cgroup, we don't need to control resources at the thread level. All processes, such as shim, qemu, vitiofsd, are able to be placed in the pod cgroup.

Reply:
> why is this only done in threaded mode?

The threaded mode is enabled when cgroup v2 is enabled and `sandbox_cgroup_only` is disabled. Otherwise, either we are using cgroup v1, or we don't expect the overhead cgroup.

- Cgroup v1 doesn't support threaded mode;
- `sandbox_cgroup_only = true`: All processes will be placed into the sandbox cgroup, so we don't need to use threaded mode.

Reply:
My question was: why do you only enable the controllers for threaded mode?

`new_cgroup` is called for overhead cgroup and sandbox cgroup:
```
        let sandbox_cgroup = new_cgroup(&config.path).context("new sandbox cgroup")?;
        let overhead_cgroup = if config.sandbox_cgroup_only {
            None
        } else {
            Some(new_cgroup(&config.overhead_path).context("new overhead cgroup")?)
        };
```
So:
* why do you not enable the controllers for cgroupv1?
* why do you not want to enable the controllers when `sandbox_cgroup_only=false`?

Reply:
If don't call `set_specified_controllers()`, then all supported controllers will be enabled, please see [1].

1: https://github.com/kata-containers/cgroups-rs/blob/main/src/cgroup.rs#L529